validator/sawtooth_validator/journal/publisher_ce.py

```diff
@@ -0,0 +1,848 @@
+                try:
+                    batch = self._batch_queue.get(timeout=0.1)
+                    self._block_publisher.on_batch_received(batch)
+                except queue.Empty:
+                    # If getting a batch times out, just try again.
+                    pass
+
+                if self._exit:
+                    return
+
+        # pylint: disable=broad-except
+        except Exception as exc:
+            LOGGER.exception(exc)
+            LOGGER.critical("BlockPublisher thread exited with error.")
+
+    def stop(self):
+        self._exit = True
+
+
+class _CandidateBlock(object):
+    """This is a helper class for the BlockPublisher. The _CandidateBlock
+    tracks all the state associated with the Block that is being built.
+    This allows the BlockPublisher to focus on when to create and finalize
+    a block and not worry about how the block is built.
+    """
+
+    def __init__(self,
+                 block_store,
+                 scheduler,
+                 committed_txn_cache,
+                 block_builder,
+                 max_batches,
+                 batch_injectors,
+                 settings_view,
+                 identity_signer,
+                 ):
+        self._pending_batches = []
+        self._pending_batch_ids = set()
+        self._injected_batch_ids = set()
+        self._block_store = block_store
+        self._scheduler = scheduler
+        self._committed_txn_cache = committed_txn_cache
+        # Look-up cache for transactions that are committed in the current
+        # chain and the state of the transactions already added to the
+        # candidate block.
+        self._block_builder = block_builder
+        self._max_batches = max_batches
+        self._batch_injectors = batch_injectors
+
+        self._settings_view = settings_view
+        self._identity_signer = identity_signer
+
+    def __del__(self):
+        self.cancel()
+
+    def cancel(self):
+        self._scheduler.cancel()
+
+    @property
+    def previous_block_id(self):
+        return self._block_builder.previous_block_id
+
+    def _last_batch(self):
+        if self._pending_batches:
+            return self._pending_batches[-1]
+        raise ValueError(
+            'last_batch called on an empty block.'
+            'Empty block publishing is not supported.'
+        )
+
+    @property
+    def batches(self):
+        return self._pending_batches.copy()
+
+    def can_add_batch(self):
+        return (
+            self._max_batches == 0
+            or len(self._pending_batches) < self._max_batches
+        )
+
+    def _check_batch_dependencies(self, batch, committed_txn_cache):
+        """Check the dependencies for all transactions in this are present.
+        If all are present the committed_txn is updated with all txn in this
+        batch and True is returned. If they are not return failure and the
+        committed_txn is not updated.
+        :param batch: the batch to validate
```

Comment:
Are we keeping these docstrings in rst format because the rest of the journal/ files are also?